### PR TITLE
5.x: Check packages comparison to YAML files

### PR DESCRIPTION
**Pull request by krzywon**
This is the same work done in #1482 but against ESS_GUI rather than master. The only difference are the yaml files I'm pointing to. I used the strictest yamls, rather than the 'min' versions.

### CODE REVIEW

**Comment by wpotrzebowski** [comment]
5.0 ready for testing on OSX

File: check_packages.py [hunk 1]
```diff
@@ -1,99 +1,94 @@
 """
-Checking and reinstalling the external packages
+Check the existing environment against expected values
 """
 from __future__ import print_function
 
 import sys
+import yaml
```
Comment (rozyczko) [inline check_packages.py]: There should be a check whether the current environment has `yaml` installed. If not, prompt the user to install it, preferably showing the exact command to do so.

**Comment by pkienzle** [comment]
Works for my on py36 after `conda install pyyaml colorama`.

If not using anaconda then `pip install pyyaml colorama` will also work.

**Comment by smk78** [comment]
Not working for me in my Anaconda3/Win10/x64/Py3.6.7 even after installing pyyaml & colorama.

```
Common Packages
setuptools Version Installed: 45.2.0.post20200209
pyparsing Version Installed: 2.4.6
html5lib Version Installed: 1.0.1
reportlab Version Installed: 3.5.34
h5py Version Installed: 2.9.0
lxml Version Installed: 4.3.0.0
Traceback (most recent call last):
  File "check_packages.py", line 62, in <module>
    print("%s Version Installed: %s"% (package_name, getattr(i, test_vals['test'])))
AttributeError: module 'PIL.Image' has no attribute 'VERSION'
```

**Comment by smk78** [comment]
Have deleted my earlier comment about pyyaml not fixing it for me in Py3; I was in the wrong branch! Sorry! check_packages _also_ works for me in Anaconda3 once I installed pyyaml & colorama.

**Comment by wpotrzebowski** [comment]
Indeed, the script runs when packages are installed according to instructions. Should we add pyyaml and colorama to the enviroment yaml file?

After running the script I got an error:
Traceback (most recent call last):
  File "check_packages.py", line 81, in <module>
    package_version = text_split[1]
IndexError: list index out of range

**Comment by rozyczko** [comment]
I get the same error as Wojtek - `index out of range`

**Comment by smk78** [comment]
Just checked again. This was before doing a pull (so the branch as of Sat):

```
(qt5) D:\SasViewCodeDev\ticket1487-5x-yaml>python check_packages.py

[0m....COMPARING PACKAGES LISTED IN ./build_tools/conda_qt5_win.yml to INSTALLED PACKAGES....

[0mmatplotlib - Expected Version Installed: 2.2.3
[0mscipy - Expected Version Installed: 1.2.0
[0mhdf5 - Expected Version Installed: 1.10.4
[0mh5py - Expected Version Installed: 2.9.0
[0mipython - Expected Version Installed: 6.5.0
[0mipykernel - Expected Version Installed: 4.9.0
[0mjupyter - Expected Version Installed: 1.0.0
[0mtwisted - Expected Version Installed: 18.9.0
[0mlxml - Expected Version Installed: 4.3.0
[0msphinx - Expected Version Installed: 1.8.4
[0mcython - Expected Version Installed: 0.29.4
[0mpytest - Expected Version Installed: 4.2.0
[0mmako - Expected Version Installed: 1.0.7
[0mpyinstaller - Expected Version Installed: 3.3.1
[0mpyopencl - Expected Version Installed: 2018.2.2
[0mtestpath - Expected Version Installed: 0.3.1
[0m
```

And this was after pulling in Jeff's two commits from yesterday:

```
(qt5) D:\SasViewCodeDev\ticket1487-5x-yaml>python check_packages.py

[0m....COMPARING PACKAGES LISTED IN ./build_tools/conda_qt5_win.yml to INSTALLED PACKAGES....

[0mmatplotlib - Expected Version Installed: 2.2.3
[0mscipy - Expected Version Installed: 1.2.0
[0mhdf5 - Expected Version Installed: 1.10.4
[0mh5py - Expected Version Installed: 2.9.0
[0mipython - Expected Version Installed: 6.5.0
[0mipykernel - Expected Version Installed: 4.9.0
[0mjupyter - Expected Version Installed: 1.0.0
[0mtwisted - Expected Version Installed: 18.9.0
[0mlxml - Expected Version Installed: 4.3.0
[0msphinx - Expected Version Installed: 1.8.4
[0mcython - Expected Version Installed: 0.29.4
[0mpytest - Expected Version Installed: 4.2.0
[0mmako - Expected Version Installed: 1.0.7
[0mpyinstaller - Expected Version Installed: 3.3.1
[0mpyopencl - Expected Version Installed: 2018.2.2
[0mtestpath - Expected Version Installed: 0.3.1
[0m
```

I do note it looks like it's stopping prematurely?

**Comment by pkienzle** [comment]
No, not stopping prematurely.  The last line in Jeff's script is the `<ESC>[0m` to reset the colors to normal for the terminal.

The index problem for ubuntu comes because many of the items do not specify a version number (dbus, gst-plugins-base, gstreamer, libtiff, sip, sqlite).  

There will also be problems with those packages which specify a minimum or maximum version (six>=1.11.0,  qt>=5.9, pyqt>=5.9.0).

**Comment by krzywon** [comment]
Good catch, Paul. I will need to do a bit better at checking versions.

File: check_packages.py [hunk 2]
```diff
@@ -1,99 +1,124 @@
...
-        i = __import__(test_vals['import_name'], fromlist=[''])
-        if test_vals['test'] is None:
-            print("%s Installed (Unknown version)" % package_name)
-        elif package_name == 'lxml':
-            verstring = str(getattr(i, 'LXML_VERSION'))
-            print("%s Version Installed: %s"% (package_name, verstring.replace(', ', '.').lstrip('(').rstrip(')')))
+        yaml_dict = yaml.load(stream, Loader=yaml.SafeLoader)
+        yaml_packages = yaml_dict['dependencies']
+        common_required_package_list = []
+        pip_packages = yaml_packages.pop()
```
Comment (pkienzle) [inline check_packages.py]: Trusting that 'pip' is the last entry in the list is a bit risky.  yaml will probably allow additional packages to be included after the pip package, but the process for generating the yaml file probably won't.

Maybe:
```python
inter_package_list = [p for entry in yaml_packages for p in (entry['pip'] if isinstance(entry, dict) else [entry])]
common_required_package_list = [re.sub("[<>=]", "=", package) for package in inter_package_list]  
```
or using for-loops:
```python
inter_package_list = []
for package in yaml_packages:
    if isinstance(package, dict):
        assert "pip" in package, f"yaml dependencies can only handle pip sub-lists not {package}"
        inter_package_list.extend(package['pip'])
    else:
        inter_package_list.append(package)
common_required_package_list = [re.sub("[<>=]", "=", package) for package in inter_package_list]  
```

**Comment by butlerpd** [comment]
so can we now merge this pull request? .. and move the kanban board card to the done pile?

**Pull request merged**